numpy/_core/src/multiarray/getset.c

```diff
@@ -370,6 +378,21 @@ array_nbytes_get(PyArrayObject *self, void *NPY_UNUSED(ignored))
 static int
 array_descr_set(PyArrayObject *self, PyObject *arg, void *NPY_UNUSED(ignored))
 {
+    // to be replaced with PyUnstable_Object_IsUniquelyReferenced https://github.com/python/cpython/pull/133144
+    int unique_reference = (Py_REFCNT(self) == 1);
+
+    if (!unique_reference) {
+        // this will not emit deprecation warnings for all cases, but for most it will
+        /* DEPRECATED 2025-05-04, NumPy 2.3 */
+        int ret = PyErr_WarnEx(PyExc_DeprecationWarning,
+                    "Setting the dtype on a Numpy array has been deprecated in Numpy 2.3.\n"
+                    "Instead of changing the dtype on an array x, create a new array with x.view(new_dtype)",
```

Comment:
```suggestion
                    "Setting the dtype on a NumPy array has been deprecated in NumPy 2.3.\n"
                    "Instead of changing the dtype on an array x, create a new array with x.view(new_dtype)",
```